src/app/qgsmaptoolchamferfillet.cpp

```diff
@@ -0,0 +1,615 @@
+const QgsSettingsEntryDouble *QgsMapToolChamferFillet::settingsValue1 = new QgsSettingsEntryDouble( QStringLiteral( "chamferfillet-fillet-value1" ), QgsSettingsTree::sTreeDigitizing, 0.0, QStringLiteral( "For fillet/chamfer operations, radius or distance1." ) );
+const QgsSettingsEntryDouble *QgsMapToolChamferFillet::settingsValue2 = new QgsSettingsEntryDouble( QStringLiteral( "chamferfillet-fillet-value2" ), QgsSettingsTree::sTreeDigitizing, 0.0, QStringLiteral( "For chamfer operation, distance2." ) );
+const QgsSettingsEntryBool *QgsMapToolChamferFillet::settingsLock1 = new QgsSettingsEntryBool( QStringLiteral( "chamferfillet-fillet-lock1" ), QgsSettingsTree::sTreeDigitizing, false, QStringLiteral( "For fillet/chamfer operations, locks distance1." ) );
+const QgsSettingsEntryBool *QgsMapToolChamferFillet::settingsLock2 = new QgsSettingsEntryBool( QStringLiteral( "chamferfillet-fillet-lock2" ), QgsSettingsTree::sTreeDigitizing, false, QStringLiteral( "For fillet/chamfer operations, locks distance2." ) );
+
+QgsMapToolChamferFillet::QgsMapToolChamferFillet( QgsMapCanvas *canvas )
+  : QgsMapToolEdit( canvas )
+  , mSnapIndicator( std::make_unique<QgsSnapIndicator>( canvas ) )
+{
+  mToolName = tr( "Chamfer or fillet" );
+}
+
+QgsMapToolChamferFillet::~QgsMapToolChamferFillet()
+{
+  cancel();
+}
+
+
+void QgsMapToolChamferFillet::applyOperationFromWidget( Qt::KeyboardModifiers modifiers )
+{
+  if ( mSourceLayer && !mOriginalGeometryInSourceLayerCrs.isNull() )
+  {
+    double value1 = mUserInputWidget->value1();
+    double value2 = mUserInputWidget->value2();
+    if ( !qgsDoubleNear( value1, 0 ) && !qgsDoubleNear( value2, 0 ) )
+    {
+      mGeometryModified = true;
+      applyOperation( value1, value2, modifiers );
+    }
+  }
+}
+
+void QgsMapToolChamferFillet::applyOperation( double value1, double value2, Qt::KeyboardModifiers )
+{
+  if ( !mSourceLayer || qgsDoubleNear( value1, 0.0 ) || qgsDoubleNear( value2, 0.0 ) )
+  {
+    cancel();
+    return;
+  }
+
+  updateGeometryAndRubberBand( value1, value2 );
+
+  // no modification
+  if ( !mGeometryModified )
+  {
+    cancel();
+    return;
+  }
+
+  if ( !mModifiedGeometry.isGeosValid() )
+  {
+    QString message;
+    QString lastError;
+    int i = 0;
+    for ( QgsAbstractGeometry::const_part_iterator ite = mModifiedGeometry.const_parts_begin(); ite != mModifiedGeometry.const_parts_end(); ite++ )
+    {
+      if ( !( *ite )->isValid( lastError ) )
+      {
+        message += QStringLiteral( "Invalid part %1: '%2'" ).arg( i ).arg( lastError );
+        break;
+      }
+      i++;
+    }
+
+    emit messageEmitted( tr( "Generated geometry is not valid: '%1'. " ).arg( mModifiedGeometry.lastError() ) + message, Qgis::MessageLevel::Critical );
```

Comment:
This is going to generate quite a big error message. Better add it to the log rather than the messagebar?